zerver/lib/notifications.py

```diff
@@ -298,10 +299,10 @@ def handle_missedmessage_emails(user_profile_id, missed_email_events):
     if not receives_offline_notifications(user_profile):
         return
 
-    messages = [um.message for um in UserMessage.objects.filter(user_profile=user_profile,
-                                                                message__id__in=message_ids,
-                                                                flags=~UserMessage.flags.read)]
-    if not messages:
+    messages = Message.objects.filter(usermessage__user_profile_id=user_profile,
+                                      id__in=message_ids,
+                                      usermessage__flags=~UserMessage.flags.read)
+    if not messages.exists():
         return
```

Comment:
But don't we need to fetch the messages anyway later on, so we end up doing a database query for the boolean check and then a second to fetch the message IDs (if any were returned)?  Whereas previously we did a single query unconditionally.

It is a cheaper query in the case that there are no remaining unread messages in the list, but that case is rare.